op-node/rollup/derive/batch_queue.go

```diff
@@ -111,43 +147,70 @@ func (bq *BatchQueue) NextBatch(ctx context.Context, safeL2Head eth.L2BlockRef)
 	} else if err != nil {
 		return nil, err
 	}
-	return batch, nil
+
+	var nextBatch *SingularBatch
+	switch batch.GetBatchType() {
+	case SingularBatchType:
+		singularBatch, ok := batch.(*SingularBatch)
+		if !ok {
+			return nil, NewCriticalError(errors.New("failed type assertion to SingularBatch"))
+		}
+		nextBatch = singularBatch
+	case SpanBatchType:
+		spanBatch, ok := batch.(*SpanBatch)
+		if !ok {
+			return nil, NewCriticalError(errors.New("failed type assertion to SpanBatch"))
+		}
+		// If next batch is SpanBatch, convert it to SingularBatches.
+		singularBatches, err := spanBatch.GetSingularBatches(bq.l1Blocks, safeL2Head)
+		if err != nil {
+			return nil, NewCriticalError(err)
+		}
+		bq.nextSpan = singularBatches
+		nextBatch = bq.popNextBatch(safeL2Head)
+	default:
+		return nil, NewCriticalError(fmt.Errorf("unrecognized batch type: %d", batch.GetBatchType()))
+	}
+
+	bq.advanceEpochMaybe(nextBatch)
```

Comment:
Why are we attempting to advance the epoch here?

Reply:
Attempting to advance the epoch is moved to here [from `deriveNextBatch`](https://github.com/ethereum-optimism/optimism/pull/7621/files/0b719a9678a71204eaa82c9642042d98697b00c4#diff-70e161ad6625124a29f49468d591c738ddec2e962fcc198d3a9db85b242c1096L212-L215), because the epoch can be advanced at the middle of a single span batch. We should attempt to advance the epoch for every singular batch.